leader-election-impl/src/main/java/com/palantir/paxos/PaxosQuorumChecker.java

```diff
@@ -189,90 +173,50 @@ private PaxosQuorumChecker() {
-                        break;
-                    }
-
                     Future<RESPONSE> responseFuture = responseCompletionService.poll(
                             deadline - System.nanoTime(),
                             TimeUnit.NANOSECONDS);
-                    // check if out of responses (no quorum failure)
-                    if (responseFuture == null) {
-                        return receivedResponses;
-                    }
-
-                    // reject invalid or repeat promises
-                    RESPONSE response = responseFuture.get();
-                    if (response.isSuccessful()) {
-                        acksRecieved++;
-                    } else {
-                        nacksRecieved++;
+                    if (timedOut(responseFuture)) {
+                        break;
                     }
-
-                    // record response
-                    receivedResponses.add(response);
-                } catch (InterruptedException e) {
-                    log.warn("paxos request interrupted", e);
-                    interrupted = true;
-                    break;
+                    receivedResponses.add(responseFuture.get());
                 } catch (ExecutionException e) {
-                    nacksRecieved++;
-                    if (onlyLogOnQuorumFailure) {
-                        toLog.add(e.getCause());
-                    } else {
-                        log.warn(PAXOS_MESSAGE_ERROR, e.getCause());
-                    }
+                    receivedResponses.markFailure();
+                    encounteredErrors.add(e.getCause());
                 }
             }
-
-            // poll for extra completed futures
-            Future<RESPONSE> future;
-            while ((future = responseCompletionService.poll()) != null) {
-                try {
-                    receivedResponses.add(future.get());
-                } catch (InterruptedException e) {
-                    log.warn("paxos request interrupted", e);
-                    interrupted = true;
-                    break;
-                } catch (ExecutionException e) {
-                    log.warn(PAXOS_MESSAGE_ERROR, e.getCause());
-                }
-            }
-
+        } catch (InterruptedException e) {
+            log.warn("paxos request interrupted", e);
+            interrupted = true;
         } finally {
-            // cancel pending futures
             cancelOutstandingRequestsAfterTimeout(allFutures);
 
-            // reset interrupted flag
             if (interrupted) {
                 Thread.currentThread().interrupt();
             }
 
-            if (onlyLogOnQuorumFailure && acksRecieved < quorumSize) {
-                for (Throwable throwable : toLog) {
-                    log.warn(PAXOS_MESSAGE_ERROR, throwable);
-                }
+            if (receivedResponses.hasQuorum()) {
+                encounteredErrors.forEach(throwable -> log.warn(PAXOS_MESSAGE_ERROR, throwable));
```

Comment:
I think we will be logging every failure even if we reach quorum. If one timelock node is in a bad shape, this would cause very large number of warns in service.log right?

Reply:
shouldn't this be `!receivedResponses.hasQuorum()` ?